assert isinstance(lineno, int) errors when collecting tests (2.8 regression)

When cloning https://github.com/osiell/odoorpc/ and running `py.test --collect-only` I get:

```
=================================== ERRORS ===================================
_________________ ERROR collecting odoorpc/tests/test_env.py _________________
.venv/lib/python3.4/site-packages/_pytest/runner.py:149: in __init__
    self.result = func()
.venv/lib/python3.4/site-packages/_pytest/main.py:431: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
.venv/lib/python3.4/site-packages/_pytest/main.py:311: in _memoizedcall
    res = function()
.venv/lib/python3.4/site-packages/_pytest/main.py:431: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
.venv/lib/python3.4/site-packages/_pytest/python.py:600: in collect
    return super(Module, self).collect()
.venv/lib/python3.4/site-packages/_pytest/python.py:460: in collect
    l.sort(key=lambda item: item.reportinfo()[:2])
.venv/lib/python3.4/site-packages/_pytest/python.py:460: in <lambda>
    l.sort(key=lambda item: item.reportinfo()[:2])
.venv/lib/python3.4/site-packages/_pytest/python.py:396: in reportinfo
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c4f2840>, int)
___________ ERROR collecting odoorpc/tests/test_field_many2many.py ___________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c51af28>, int)
___________ ERROR collecting odoorpc/tests/test_field_many2one.py ____________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c3738c8>, int)
___________ ERROR collecting odoorpc/tests/test_field_one2many.py ____________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c38b268>, int)
___________ ERROR collecting odoorpc/tests/test_field_reference.py ___________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c38bea0>, int)
________________ ERROR collecting odoorpc/tests/test_login.py ________________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c385048>, int)
________________ ERROR collecting odoorpc/tests/test_model.py ________________
[...]
    assert isinstance(lineno, int)
E   assert isinstance(<function MetaModel.__getattr__.<locals>.rpc_method at 0x7fad5c385b70>, int)
```

This looks similar to #1035 which was fixed by @nicoddemus, and only happened for Python 2.6 - this is with Python 3.4.3.

It works fine with pytest 2.7.3, and when reverting 3497aa0766ac6f2f2cd66cc483c86f2a67a738aa from #921 the issue goes away as well. @tomviner

Looking at the repr, it seems it gets [this function](https://github.com/osiell/odoorpc/blob/master/odoorpc/models.py#L66) instead of a line number?

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -405,7 +405,10 @@ class PyCollector(PyobjMixin, pytest.Collector):
         """ Look for the __test__ attribute, which is applied by the
         @nose.tools.istest decorator
         """
-        return safe_getattr(obj, '__test__', False)
+        # We explicitly check for "is True" here to not mistakenly treat
+        # classes with a custom __getattr__ returning something truthy (like a
+        # function) as test classes.
+        return safe_getattr(obj, '__test__', False) is True
 
     def classnamefilter(self, name):
         return self._matches_prefix_or_glob_option('python_classes', name)

--- a/testing/python/integration.py
+++ b/testing/python/integration.py
@@ -283,6 +283,35 @@ class TestNoselikeTestAttribute:
         assert len(call.items) == 1
         assert call.items[0].cls.__name__ == "TC"
 
+    def test_class_with_nasty_getattr(self, testdir):
+        """Make sure we handle classes with a custom nasty __getattr__ right.
+
+        With a custom __getattr__ which e.g. returns a function (like with a
+        RPC wrapper), we shouldn't assume this meant "__test__ = True".
+        """
+        # https://github.com/pytest-dev/pytest/issues/1204
+        testdir.makepyfile("""
+            class MetaModel(type):
+
+                def __getattr__(cls, key):
+                    return lambda: None
+
+
+            BaseModel = MetaModel('Model', (), {})
+
+
+            class Model(BaseModel):
+
+                __metaclass__ = MetaModel
+
+                def test_blah(self):
+                    pass
+        """)
+        reprec = testdir.inline_run()
+        assert not reprec.getfailedcollections()
+        call = reprec.getcalls("pytest_collection_modifyitems")[0]
+        assert not call.items
+
 
 @pytest.mark.issue351
 class TestParameterize:

